### PR TITLE
Fix alias listing

### LINKED ISSUE #105: AliasRepository::loadWithMailbox() has wrong like-check
If you have a mailbox called lastname@example.com and an alias
something@example.com -> admin@example.com, firstname.lastname@example.com

AliasRepository::loadWithMailbox() will also include this alias although it is completely unrelated. This causes a problem when purging mailboxes, as they try to delete this alias too.

The SQL like query should be replaced to something like find_in_set() or, if that is not doable due to other DBs not supporting it, one can choose
`a.goto like ',' & ?2 OR a.goto like ?2 & ','`

### PR DESCRIPTION
**Pull request by liayn**
Fixes #105

### CODE REVIEW

**Comment by liayn** [comment]
@barryo Just to repeat my message here: The change is NOT tested. Will try to do so this week.

**Comment by barryo** [comment]
Understood 😄 

I'll sit on it until you come back (or if I get a chance to look and test myself in the meantime).

**Comment by PhrozenByte** [comment]
Without testing anything: The query probably lacks `a.goto = ?2` (i.e. a single goto address).

File: application/Repositories/Alias.php [hunk 1]
```diff
@@ -61,14 +61,20 @@ public function loadForMailbox( $mailbox, $admin, $ima = false )
...
 	{
-		$qb = $this->getEntityManager()->createQueryBuilder()
-                ->select( 'a' )
-                ->from( '\\Entities\\Alias', 'a' )
-                ->where( 'a.address != a.goto' )
-                ->andWhere( 'a.goto != ?1' )
-                ->andWhere( 'a.goto like ?2' )
-                ->setParameter( 1, $mailbox->getUsername() )
-                ->setParameter( 2, '%' . $mailbox->getUsername() . '%');
+		$qb = $this->getEntityManager()->createQueryBuilder();
```
Comment (liayn) [inline application/Repositories/Alias.php]: indentation wrong

**Comment by liayn** [comment]
Will think the whole thing through again.

**Comment by PhrozenByte** [comment]
Still without testing anything: I think the current solution doesn't actually solve the problem in #105, the `LIKE` condition still matches `firstname.lastname@example.com` when searching for `lastname@example.com`.

The query should look like the following (i.e. check first item, intermediate items and last item):
```php
        $qb = $this->getEntityManager()->createQueryBuilder()
            ->select( 'a' )
            ->from( '\\Entities\\Alias', 'a' )
            ->where($qb->expr()->andX(
                'a.address != a.goto',
                'a.goto != ?1',
                $qb->expr()->orX(
                    'a.goto like ?2',
                    'a.goto like ?3',
                    'a.goto like ?4'
                )
            ))
            ->setParameter( 1, $mailbox->getUsername() )
            ->setParameter( 2, $mailbox->getUsername() . ',%')
            ->setParameter( 3, '%,' . $mailbox->getUsername() . ',%')
            ->setParameter( 3, '%,' . $mailbox->getUsername());
```

String concatenation is a pain in the ass with SQL... Avoid it whenever possible. I'm not sure whether Doctrine rewrites queries to support it anyway (it probably does; [Doctrine suggests the `CONCAT()` function](http://docs.doctrine-project.org/projects/doctrine-orm/en/latest/reference/dql-doctrine-query-language.html#dql-functions)), but the syntax chaos is an absolute mess!

The ISO/ANSI SQL operator is `||`, however, many DBMS, including MySQL (ok, this isn't fully true, it does support it by setting a compatibility flag...), don't support it. For others, like SQLite and Sybase, it's the only supported operator. Microsoft is always a highlight: Microsoft Access supports the `&` operator (what is usually the binary/bitwise operator...), Microsoft SQL Server 2008 the `+` operator. Most DBMS don't just support a single operator... The probably most supported way of concatenating strings is the `CONCAT()` function: it is supported by MySQL, MariaDB, PostgreSQL, Oracle, IBM DB2, IBM Informix, SAP HANA and Microsoft SQL Server 2012+. Amazing, isn't it? 😒

**Comment by liayn** [comment]
@barryo Finally found the time to test this. Works like a charm here. ;-)

**Comment by liayn** [comment]
Indentation fixed now as well for this file.

**Comment by liayn** [comment]
@barryo ping

**Comment by barryo** [comment]
Cheers @liayn 🍻

**Comment by liayn** [comment]
Cheers and thanks.

**Pull request merged**